.github/workflows/release.yaml

```diff
@@ -0,0 +1,912 @@
+        uses: actions/upload-artifact@v3
+        with:
+          if-no-files-found: error
+          retention-days: 2
+          name: artifacts-arm
+          path: |
+            ./out-${{ env.ARTIFACT }}-${{ matrix.ghc }}.tar
+
+  build-mac-x86_64:
+    name: Build binary (Mac x86_64)
+    runs-on: macOS-10.15
+    env:
+      MACOSX_DEPLOYMENT_TARGET: 10.13
+      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+      S3_HOST: ${{ secrets.S3_HOST }}
+      ADD_CABAL_ARGS: ""
+      ARTIFACT: "x86_64-apple-darwin"
+      ARCH: 64
+      TARBALL_EXT: tar.xz
+      DISTRO: na
+    strategy:
+      fail-fast: false
+      matrix:
+        ghc: ["9.4.4", "9.2.5", "9.0.2", "8.10.7"]
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+
+      - name: Run build
+        run: |
+          brew install coreutils tree
+          bash .github/scripts/build.sh
+          tar cf out-${ARTIFACT}-${GHC_VERSION}.tar out/ store/
+        env:
+          GHC_VERSION: ${{ matrix.ghc }}
+
+      - if: always()
+        name: Upload artifact
+        uses: actions/upload-artifact@v3
+        with:
+          if-no-files-found: error
+          retention-days: 2
+          name: artifacts-mac-x86_64
+          path: |
+            ./out-${{ env.ARTIFACT }}-${{ matrix.ghc }}.tar
+
+  build-mac-aarch64:
+    name: Build binary (Mac aarch64)
+    runs-on: [self-hosted, macOS, ARM64]
+    env:
+      MACOSX_DEPLOYMENT_TARGET: 10.13
+      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+      S3_HOST: ${{ secrets.S3_HOST }}
+      ADD_CABAL_ARGS: ""
+      ARTIFACT: "aarch64-apple-darwin"
+      ARCH: ARM64
+      TARBALL_EXT: tar.xz
+      DISTRO: na
+      HOMEBREW_CHANGE_ARCH_TO_ARM: 1
+    strategy:
+      fail-fast: false
+      matrix:
+        ghc: ["9.4.4", "9.2.5", "8.10.7"]
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+
+      - name: Run build
+        run: |
+          bash .github/scripts/brew.sh git coreutils llvm@11 autoconf automake tree
+          export PATH="$HOME/.brew/bin:$HOME/.brew/sbin:$HOME/.brew/opt/llvm@11/bin:$PATH"
+          export CC="$HOME/.brew/opt/llvm@11/bin/clang"
+          export CXX="$HOME/.brew/opt/llvm@11/bin/clang++"
+          export LD=ld
+          export AR="$HOME/.brew/opt/llvm@11/bin/llvm-ar"
+          export RANLIB="$HOME/.brew/opt/llvm@11/bin/llvm-ranlib"
+          bash .github/scripts/build.sh
+          tar cf out-${ARTIFACT}-${GHC_VERSION}.tar out/ store/
+        env:
+          GHC_VERSION: ${{ matrix.ghc }}
+
+      - if: always()
+        name: Upload artifact
+        uses: actions/upload-artifact@v3
+        with:
+          if-no-files-found: error
+          retention-days: 2
+          name: artifacts-mac-aarch64
+          path: |
+            ./out-${{ env.ARTIFACT }}-${{ matrix.ghc }}.tar
+
+  build-win:
```

Comment:
It seems to me that we probably could merge all the build jobs into one. The main differences that I can see are:
- Different preparatory work in setting up the environment etc. This could be solved by having a per-platform bash script that we just call to do that.
- Some different environment variables that need to be set. Again, this could be done by the platform-specfic setup script.
- Some things need a different shell, but I think we could put that in a variable

That would make this a _lot_ easier to maintain, the hundreds of lines of very similar (but slightly different in inessential ways) yaml is going to be a huge pain.

Reply:
> It seems to me that we probably could merge all the build jobs into one. The main differences that I can see are

Strongly disagree. The steps are wildly different and so are the matrices. The linux jobs have matrices per distro, the others don't. I'm against unifying this, it will be a huge mess.